Useful Stethoscopes and Penlights

LovliestPlant opened the pull request:
## About The Pull Request
This PR overhauls and adds functionality to Stethoscopes and Flashlight examinations (i.e. shining a penlight into someone's eyes/mouth).

Stethoscope Changes
- Dynamic results based on targeted body part.
- Chest: Assess whether the heart is beating, whether it is damaged (above 10 dmg), and whether or not the player has suffered severe blood loss. Assess whether the player is breathing, has lung damage (above 10dmg), and whether or not they have suffocation damage (above 10dmg).
- Head/Extremities: Assess heart function as above, and blood level (nominal/low/critical).
- Groin: Assess whether or not the liver or the appendix have suffered damage (above 10dmg). (The latter is only detectable if the patient is conscious)

Flashlight Changes
- Eyes: Assess whether the patient is dead, has sustained brain damage (above 20dmg), or has an X-Ray mutation.
- Mouth: Assess mouth organs, dental implants, suffocation damage (above 20dmg), and blood volume (nominal/low/critical)
## Why It's Good For The Game
Adds some proper functionality to two very rarely used tools.

Loosens up medical players' dependence on health analyzers to assess their patients Provides an alternative, if manual and less accurate, means of assessment.

For players interested in a more hands-on, RP focused approach to medbay (or those in areas without power to recharge analyzers); I think this will be a welcome addition.
## Proof of Testing


<details>
<summary>Screenshots/Videos</summary>

https://github.com/tgstation/tgstation/assets/107971606/43e87774-6db5-4db1-b7fa-92c3565c1009


https://github.com/tgstation/tgstation/assets/107971606/9ec42cc4-a996-4d60-a6f1-1aecf24b9bc8

</details>

## Changelog
:cl:
add: Stethoscopes may be used on the chest, groin, or extremities to assess organ damage, blood level, and/or suffocation damage depending on the targeted area.
add: Shining flashlights into the mouth or eyes of other players will additionally assess brain health, suffocation damage, and/or blood level depending on the targeted area.
balance: Halves the duration of the flash effect from shining lights into players' eyes (2s -> 1s). Use combat mode to get the full duration.
/:cl:

LovliestPlant committed:
The Goods

- Overhauls functionality of stethoscopes
- Overhauls information provided by flashlight exams

Wallemations commented:
Why would you need a tool to tell if someone's dead when you can just shift+right-click to tell that? I feel like it the blindness check made way more sense than this does.
Rather than checking for death I think it should keep the blindness/eye-damage check but keep that brain damage addition you made as well.

Wallemations commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -121,23 +124,28 @@
+					render_list += "<span class='notice'>You direct [src] to [M]'s eyes:</span>\n"
+
+					if(M.stat == DEAD || !M.flash_act(visual = 1))
+						render_list += "<span class='danger ml-1'>[M.p_their(TRUE)] pupils don't react to the light!</span>\n"//mob is dead
```
Why are you using `<span class='n'>` instead of `span_[n]()`?

LovliestPlant replied:
to get the indentation in the readout! Also that's how it was done in the health analyzer, which this is largely based off of. Is there a particular advantage to one over the other?

MrMelbert replied:
You shuold be using `span_notice("blah blah")` over `<span class='notice'>Blah Blah</span>`, it's just more readable. 
I don't think we have a wrapper for the alert messages though.

LovliestPlant replied:
makes sense. How would you get the formatting in? The "ml-1"s to indent certain bits of text.

san7890 replied:
you can't do the `ml-1` stuff here sadly because it's a limitation of that framework. should definitely do that for stuff like on line 128.

LovliestPlant replied:
ah okay. Cool, I'll do a quick readability pass then.

Y0SH1M4S73R commented:
> Why would you need a tool to tell if someone's dead when you can just shift+right-click to tell that?

Because of `TRAIT_FAKEDEATH`, which makes them show up as dead on examining, medhuds, the kiosk, and non-advanced health analyzers. This flashlight test on the eyes only checks `stat`, so it sees through fake death such as through zombie powder or undeath.

HoodMilk commented:
Does the penlight still blind people?

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -121,23 +124,28 @@
+					user.visible_message(span_warning("[user] directs [src] to [M]'s eyes."), ignored_mobs = user)
+					render_list += "<span class='notice'>You direct [src] to [M]'s eyes:</span>\n"
+
+					if(M.stat == DEAD || !M.flash_act(visual = 1))
```
Your pr body says `Removes ability to detect blindness with this examination.`, but you still have the ability to detect blindness right there. Flash act returns false if it fails (they are blind).

MrMelbert replied:
Weird facet about the stat check as well - you are blind when unconscious or in hard crit. So it'll fail on mobs which are unconscious, hard crit, in fake death (regen coma / ling stasis), or dead.

LovliestPlant replied:
(I didn't actually know what flash_act was, and didn't want to break anything ;-;)
Nixed that check and cleaned up the code, should function as advertised now! (and no longer produce a screen flash effect)

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -173,27 +181,62 @@
+
+				//assess any suffocation damage
+				var/mob/living/carbon/carbon_patient = M
+				var/oxy_loss = carbon_patient.getOxyLoss()
+				var/hypoxia_status = FALSE
+
+				if(oxy_loss > 20)//target has suffocation damage
+					hypoxia_status = TRUE
```
```suggestion
				var/hypoxia_status = carbon_patient.getOxyLoss() > 20
```
As far as I can tell you don't really use anything else, might as well skip the middle man

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -173,27 +181,62 @@
+						render_list += "<span class='notice ml-1'>[M] has [pill_count] pill[pill_count > 1 ? "s" : ""] implanted in [ M.p_their()] teeth.</span>\n"
+
+				//assess any suffocation damage
+				var/mob/living/carbon/carbon_patient = M
```
`M` is already cast to carbon.. why do this?

MrMelbert commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,91 @@
 
 	var/obj/item/organ/internal/heart/heart = carbon_patient.get_organ_slot(ORGAN_SLOT_HEART)
 	var/obj/item/organ/internal/lungs/lungs = carbon_patient.get_organ_slot(ORGAN_SLOT_LUNGS)
+	var/obj/item/organ/internal/lungs/liver = carbon_patient.get_organ_slot(ORGAN_SLOT_LIVER)
+	var/obj/item/organ/internal/lungs/appendix = carbon_patient.get_organ_slot(ORGAN_SLOT_APPENDIX)
```
```suggestion
	var/obj/item/organ/internal/liver/liver = carbon_patient.get_organ_slot(ORGAN_SLOT_LIVER)
	var/obj/item/organ/internal/appendix/appendix = carbon_patient.get_organ_slot(ORGAN_SLOT_APPENDIX)
```

MrMelbert commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,91 @@
+
+	//assess abdominal organs
+	if(body_part == BODY_ZONE_PRECISE_GROIN)
+		if(liver.damage > 10 || appendix.damage > 10)//if there's damage to either their liver or appendix
```
Aaaand same here, sanity check for liver and appendix

LovliestPlant replied:
Reshuffled their section a bit to make it work, but added!

LovliestPlant committed:
Update code/game/objects/items/devices/flashlight.dm

combines hypoxia status determination into var declaration

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

LovliestPlant committed:
Update code/modules/clothing/neck/_neck.dm

fixes object paths for liver and appendix declarations

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

LovliestPlant committed:
Review changes

- adds sanity checks for organs existing
- adds TRAIT_NOBREATH to the reasons a patient won't, in fact, be breathing
- removes a duplicate var cast (M being cast to carbon twice)
- removes a loitering blindness check
- flashlights will no longer cause a screen flash effect

LovliestPlant commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,91 @@
+	if(body_part == BODY_ZONE_CHEST)//Check if we can actually hear lung sounds
+		if(carbon_patient.stat == DEAD || (HAS_TRAIT(carbon_patient, TRAIT_FAKEDEATH)) || carbon_patient.failed_last_breath || carbon_patient.losebreath)//If pt is dead or otherwise not breathing
+			render_list += "<span class='danger ml-1'>[M.p_theyre(TRUE)] not breathing!</span>\n"
+		else if(lungs.damage > 10)//if breathing, check for lung damage
```
added!

LovliestPlant commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,91 @@
+	//assess breathing
+	if(body_part == BODY_ZONE_CHEST)//Check if we can actually hear lung sounds
+		if(carbon_patient.stat == DEAD || (HAS_TRAIT(carbon_patient, TRAIT_FAKEDEATH)) || carbon_patient.failed_last_breath || carbon_patient.losebreath)//If pt is dead or otherwise not breathing
+			render_list += "<span class='danger ml-1'>[M.p_theyre(TRUE)] not breathing!</span>\n"
```
That is. A very good point, thank u. Added!

LovliestPlant commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,91 @@
+
+	//assess heart & blood level
+	if(body_part == BODY_ZONE_CHEST)//if we're listening to the chest
+		if(!heart.beating || carbon_patient.stat == DEAD)
```
Added!

MrMelbert commented:
... Why remove the flash effect? You are shining a light into (yours/someone's) eyes. Why would it not flash / blind them momentarily?

LovliestPlant commented:
> ... Why remove the flash effect? You are shining a light into (yours/someone's) eyes. Why would it not flash / blind them momentarily?

The idea is that should this be done with any degree of regularity, it'd get irritating for patients to deal with. Imagine how much of a headache going to medbay would be if your screen flashed white every time you were scanned with a health analyzer. It'd get old real quick, and discourage players from using the method. There's no mechanical benefit for or tied to the effect, so I figured I'd axe it for sanity down the line.

Jolly-66 commented:
I've "toyed" with a penlight in my A&P lab this past semester and they are, indeed, very fucking bright.

Keep the flash effect, or move it to an alt click function or something.

LovliestPlant committed:
re-implements flash effect

- Trims the effect duration down from 25 to 10. Noticeable but should still avoid player fatigue.
- cleans up some formatting: replaces a 'notice' with 'info' for consistency
- adds examine_block text for shining a light into your own eyes.

LovliestPlant commented:
Still a bit worried about player fatigue from the flash. But I found out the effect duration can be shortened, so it should be fine. Current duration is around half of the normal time.

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,55 @@
 					to_chat(user, span_warning("[M] doesn't have any eyes!"))
 					return
 
+				M.flash_act(length = 10)//breifly apply flash affect to target
```
Be sure to include `visual = TRUE` as it did in the past, so it doesn't actually stun people

LovliestPlant replied:
so THAT'S what that does. Added!

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,55 @@
 					to_chat(user, span_warning("[M] doesn't have any eyes!"))
 					return
 
+				M.flash_act(length = 10)//breifly apply flash affect to target
```
```suggestion
				M.flash_act(length = 1 SECONDS)//breifly apply flash affect to target
```

MrMelbert commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,55 @@
+					user.visible_message(span_warning("[user] directs [src] to [M]'s eyes."), ignored_mobs = user)
+					render_list += "<span class='info'>You direct [src] to [M]'s eyes:</span>\n"
+
+					if(M.stat == DEAD)
+						render_list += "<span class='danger ml-1'>[M.p_their(TRUE)] pupils don't react to the light!</span>\n"//mob is dead
```
Re-include the blind check here as well, since you're moving the flash act out of here

LovliestPlant replied:
good catch, added.

LovliestPlant committed:
Update code/game/objects/items/devices/flashlight.dm

swapped the vague duration of '10' for the much more informative '1 SECONDS'

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

LovliestPlant committed:
flash effect cleanup

- Sets applied flash to be visual only
- Re-adds blindness check to eyes assessment

HoodMilk commented:
The flash effect is useful for combat, and one of the only non-lethal options doctors have aside from chems.

ExcessiveUseOfCobblestone commented:
if you have combat mode on the flash duration shouldnt be shortened at all.

Im a bit confused what the worry is, if it does the printout once theres no information youre getting that is going to be actively changing like damage numbers, how are you going to be fatigued by getting flashed once?

LovliestPlant committed:
readability pass

replaces applicable <span class='info'>'s with span_info( )'s for easier reading

ExcessiveUseOfCobblestone commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,55 @@
 					to_chat(user, span_warning("[M] doesn't have any eyes!"))
 					return
 
+				M.flash_act(visual = TRUE, length = 1 SECONDS)//breifly apply flash affect to target
```
this is nerfed right from default `flash_act`? If so this should check for combat mode, combat mode going back to default

LovliestPlant replied:
added!

ExcessiveUseOfCobblestone reviewed (changes requested):
Need above addressed

LovliestPlant committed:
Adds a combat mode toggle for flash duration

LovliestPlant commented:
The main concern was that it would be awkward and disruptive to tack a 2.5s pause in the middle of an otherwise active RP scene. That was mitigated by shortening the duration of the flash effect to 1s. But yes, if a player is wielding their flashlight with violent intent the normal duration would be more appropriate.

ExcessiveUseOfCobblestone commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,113 @@
+	var/render_list = list()//information will be packaged in a list for clean display to the user
+
+	//determine what specific action we're taking
+	if(body_part == BODY_ZONE_CHEST)//We're listening to their chest
+		user.visible_message(span_notice("[user] places [src] against [carbon_patient]'s [body_part] and listens attentively."), ignored_mobs = user)
+		if(!user.can_hear())
+			to_chat(user, span_notice("You place [src] against [carbon_patient]'s [body_part]. Fat load of good it does you though, since you can't hear"))
+			return
+		else
+			render_list += span_info("You place [src] against [carbon_patient]'s [body_part]:\n")
+
+
+	else if(body_part == BODY_ZONE_PRECISE_GROIN)//We're feeling their abdomen
+		render_list += span_info("You carefully press down on [carbon_patient]'s abdomen:\n")
+		user.visible_message(span_notice("[user] presses their hands against [carbon_patient]'s abdomen."), ignored_mobs = user)
+	else if(body_part == BODY_ZONE_HEAD)//We're feeling their pulse on their neck
+		render_list += span_info("You carefully press your fingers to [carbon_patient]'s neck:\n")
+		user.visible_message(span_notice("[user] presses their fingers against [carbon_patient]'s neck."), ignored_mobs = user)
+	else if(body_part != BODY_ZONE_CHEST && body_part != BODY_ZONE_PRECISE_GROIN && body_part != BODY_ZONE_PRECISE_EYES && body_part != BODY_ZONE_PRECISE_MOUTH)//We're feeling their pulse on an extremity
+		render_list += span_info("You carefully press your fingers to [carbon_patient]'s [body_part]:\n")
+		user.visible_message(span_notice("[user] presses their fingers against [carbon_patient]'s [body_part]."), ignored_mobs = user)
+	else//We're targeting their mouth or eyes. Try again.
+		balloon_alert(user, "can't do that!")
+		return
```
Switch statement instead?

https://www.byond.com/docs/ref/index.html#/proc/switch

LovliestPlant replied:
So that's how switch statements work in DM, huh. Cheers, implemented!

ExcessiveUseOfCobblestone commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,113 @@
+				render_list += "<span class='notice ml-1'>[M.p_theyre(TRUE)] breathing normally.</span>\n"//they're okay :D
+
+	//assess heart & blood level
+	if(body_part == BODY_ZONE_CHEST)//if we're listening to the chest
```
same as L224?

ExcessiveUseOfCobblestone commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,113 @@
+			else
+				render_list += "<span class='notice ml-1'>You hear a healthy heartbeat.</span>\n"//they're okay :D
+
+	if(body_part != BODY_ZONE_CHEST && body_part != BODY_ZONE_PRECISE_GROIN && body_part != BODY_ZONE_PRECISE_EYES && body_part != BODY_ZONE_PRECISE_MOUTH)//we're checking their pulse on an extremity
```
same as 216?

ExcessiveUseOfCobblestone commented on code/modules/clothing/neck/_neck.dm:
```diff
@@ -185,26 +185,113 @@
+				render_list += "<span class='notice ml-1'>[M.p_their(TRUE)] pulse is [pulse_pressure] and [heart_strength].</span>\n"
+
+	//assess abdominal organs
+	if(body_part == BODY_ZONE_PRECISE_GROIN)
```
if you make it a switch then u move it before your gigacheck on 249 u will be able to make 249 a little more readable

LovliestPlant replied:
It hadn't occurred to me that organizing the code by vital sign involved so much repetition, good catch. It's all grouped by body part in the switch now!

LovliestPlant committed:
code cleanup

- Replaces If/else chains with a switch
- Re-organizes code to be grouped by body part, rather than vital sign

Jacquerel commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,58 @@
-						M.visible_message(span_notice("[M] directs [src] to [M.p_their()] eyes."), span_notice("You wave the light in front of your eyes."))
+						render_list += "<span class='notice ml-1'>Trippy!</span>\n"
+
+					//M.visible_message(span_notice("[M] directs [src] to [M.p_their()] eyes."), span_notice("You wave the light in front of your eyes."))
```
```suggestion
```
Don't leave legacy code present but uncommented, just delete it. Git history can restore it if we need to.

Jacquerel commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,58 @@
+				else
+					M.flash_act(visual = TRUE, length = 1 SECONDS)//breifly apply flash affect to target(~1s)
+
 				if(M == user) //they're using it on themselves
```
You've had two other maintainers pass over this _without_ requiring this so I would feel bad to do it at this late stage but it'd be great if someone replaced these single letter vars with descriptive names.

LovliestPlant replied:
I would do. But 'M' seems to be a pretty standard variable used in attack calls like this. (The same variable is used in the calls governing stethoscopes too). So I think consistency would win out over descriptiveness in this case. There's loads of comments around, so hopefully it's not too obtuse to read.

Jacquerel replied:
It's only consistent because all of that code is old too, but I'm not making you do it 😅

Jacquerel commented on code/game/objects/items/devices/flashlight.dm:
```diff
@@ -104,40 +104,58 @@
 					to_chat(user, span_warning("[M] doesn't have any eyes!"))
 					return
 
+				if(user.combat_mode)
+					M.flash_act(visual = TRUE)//Channel your anger into shining the light in their eyes for the defualt duration (~2.5s)
+				else
+					M.flash_act(visual = TRUE, length = 1 SECONDS)//breifly apply flash affect to target(~1s)
```
```suggestion
				M.flash_act(visual = TRUE, length = (user.combat_mode) ? 2.5 SECONDS : 1 SECONDS)
```

Optional suggestion.

LovliestPlant replied:
For the life of me, I'd never gotten those kinds of combined if/else statements to work. But sure thing! I'll tack a comment on the end just to ensure readability.

Jacquerel reviewed (changes requested):
This looks nearly done let's get it over the finish line

LovliestPlant committed:
Update code/game/objects/items/devices/flashlight.dm

Good catch, thought I'd cleaned all those line out

Co-authored-by: Jacquerel <hnevard@gmail.com>

LovliestPlant committed:
Update code/game/objects/items/devices/flashlight.dm

Co-authored-by: Jacquerel <hnevard@gmail.com>

LovliestPlant committed:
adds a comment to Combat mode flash duration
